### PR TITLE
feat(notifications): Support querying multiple projects for the new notification setting tables 

### PR DESCRIPTION
**Pull request by snigdhas**
Support querying multiple projects for the new notification setting tables. Followup to https://github.com/getsentry/sentry/pull/55744.

### CODE REVIEW

**Review by corps — APPROVED** [review]
I like!

File: src/sentry/services/hybrid_cloud/notifications/impl.py [hunk 1]
```diff
@@ -142,6 +152,38 @@ def get_settings_for_user_by_projects(
...
+        """
+        Returns a a mapping of project scopes settings, with projects IDs
+        mapped to a map of provider to notifications setting values.
+        """
+        user = User.objects.get(id=user_id)
+        project_to_settings: MutableMapping[
+            int, MutableMapping[ExternalProviderEnum, NotificationSettingsOptionEnum]
+        ] = defaultdict(dict)
+
+        # TODO(snigdha): This is an N+1 query. We should optimize this if there's a visible performance impact.
```
Comment (scefali) [inline src/sentry/services/hybrid_cloud/notifications/impl.py:174]: @snigdhas can we make it take a list of projects as well?

File: src/sentry/services/hybrid_cloud/notifications/impl.py [hunk 2]
```diff
@@ -142,6 +152,38 @@ def get_settings_for_user_by_projects(
...
+        for project in projects:
+            providers = get_setting_providers_for_users(
+                [user_id],
+                project=project,
+                organization=project.organization,
```
Comment (scefali) [inline src/sentry/services/hybrid_cloud/notifications/impl.py:179]: @snigdhas isn't this going to limit the query to the org scope? I'd think we just want to pass in the project here

File: src/sentry/services/hybrid_cloud/notifications/impl.py (hunk 1 above)
Comment (snigdhas) [reply src/sentry/services/hybrid_cloud/notifications/impl.py:174]: sure - made that change. We're getting to a point where each helper is taking the base settings and modifying the format into some type of map which is a bit clunky

File: src/sentry/services/hybrid_cloud/notifications/impl.py (hunk 2 above)
Comment (snigdhas) [reply src/sentry/services/hybrid_cloud/notifications/impl.py]: Hmm even if we don't pass the org, we derive the org [here](https://github.com/getsentry/sentry/blob/fd14efac64442442bebc99bab0cb3589be34b37c/src/sentry/notifications/helpers.py#L704). Are you saying we should skip that and only look at project scoped settings?

Comment (scefali) [reply src/sentry/services/hybrid_cloud/notifications/impl.py]: I guess we OR the queries so it's fine!

File: src/sentry/notifications/helpers.py [hunk 3]
```diff
@@ -702,21 +695,24 @@ def get_query(
...
         )
-        if project
+        if projects
         else Q()
     )
 
+    org_id = (
+        organization.id
+        if organization
+        else (projects[0].organization.id if projects else None)  # type:ignore
```
Comment (markstory) [inline src/sentry/notifications/helpers.py:707]:
```suggestion
        else (projects[0].organization_id if projects else None)  # type:ignore
```

We don't need to load the organization record to get its id.

File: src/sentry/notifications/helpers.py [hunk 4]
```diff
@@ -745,12 +741,13 @@ def get_all_setting_providers(
...
 
-    query = get_query(recipients, project, organization)
+    projects = [project] if project else []
+    query = get_query(recipients, projects, organization)
     return NotificationSettingProvider.objects.filter(query)
 
 
 def get_all_setting_options(
-    recipients: Iterable[RpcActor | Team | RpcUser] | None = None,
+    recipients: Iterable[RpcActor | Team | User] | None = None,
```
Comment (markstory) [inline src/sentry/notifications/helpers.py:750]: Do the functions that have their signatures changed no longer run in region silo code? There won't be a way to get `User` in issue alerts or metric alert logic, only `RpcUser` is available there.

File: src/sentry/notifications/helpers.py [hunk 5]
```diff
@@ -834,9 +832,9 @@ def get_setting_options_for_users(
     if not additional_filters:
         additional_filters = Q()
 
-    users = user_service.get_many(filter={"user_ids": user_ids})
-
-    query = get_query(recipients=users, project=project, organization=organization)
+    users = User.objects.filter(id__in=user_ids)
+    projects = [project] if project else []
```
Comment (markstory) [inline src/sentry/notifications/helpers.py:836]: Having access to both `User` and `Project` will not be possible in siloed application code. I think you might need to continue using `user_service` in notification logic that could run in region logic.

**Review by JoshFerge — COMMENTED** [review]
excuse me if i'm missing something, but are there any tests for this newly added code?

**Comment by snigdhas** [comment]
Closing this in favor of https://github.com/getsentry/sentry/pull/56064. I'll add tests and make sure it's HC-compliant. Thanks folks!

**Pull request closed without merging**